r/src/array.cpp

```diff
@@ -323,6 +357,37 @@ inline SEXP StringArray_to_Vector(const std::shared_ptr<arrow::Array>& array) {
   return res;
 }
 
+SEXP DictionaryArray_to_Vector(arrow::DictionaryArray* dict_array) {
+  auto dict = dict_array->dictionary();
+  auto indices = dict_array->indices();
+
+  if (dict->type_id() != Type::STRING || indices->type_id() != Type::INT32) {
+    stop("Cannot convert Dictionary Array of type `%s`", dict_array->type()->ToString());
+  }
+
+  IntegerVector f = simple_Array_to_Vector<INTSXP>(indices);
+  f = f + 1;
+  f.attr("levels") = StringArray_to_Vector(dict);
+  if (dict_array->dict_type()->ordered()) {
+    f.attr("class") = CharacterVector::create("ordered", "factor");
+  } else {
+    f.attr("class") = "factor";
+  }
+  return f;
+}
+
+SEXP Date32Array_to_Vector(const std::shared_ptr<arrow::Array>& array) {
+  IntegerVector out(simple_Array_to_Vector<INTSXP>(array));
+  out.attr("class") = "Date";
+  return out;
+}
+
+SEXP Date64Array_to_Vector(const std::shared_ptr<arrow::Array>& array) {
+  NumericVector out(simple_Array_to_Vector<REALSXP>(array));
```

Comment:
Hm, maybe should comment about this since Date64 is an int64-based representation

Reply:
Yeah there’s no int64 in R so for now we’ll have to stick to doubles which give lossless within [-2^52, 2^52]. I’ll look into what dates these are. 

We could also warn for dates outside of the range but that feels expensive.

Reply:
So we're fine (if I understand this correctly) between these two dates: 

```r
> structure(c(-2^52, 2^52), class = c("POSIXct", "POSIXt"))
[1] "-142711421-01-25 20:21:05 LMT" "142715360-12-06 04:48:16 CET"
```

Reply:
sounds good to me

Reply:
✅